requests-for-comments/2022-06-01-session-analysis.md

```diff
@@ -0,0 +1,118 @@
+
+![image](https://user-images.githubusercontent.com/4813045/171521347-8f3fd16c-28ab-49ac-9baa-c02097d4c42c.png)
+
+Details: The “auto-bucket” functionality would require more work to figure out exactly how this should work + what the UX should be.
+
+### 5. Count of a specific event per session over time
+
+Proposal: Users accomplish this using formulas. They would make a series that is (a) Session count and (b) total count, and then divide (b) by (a).
+
+![image](https://user-images.githubusercontent.com/4813045/171521365-2b353bb0-7634-45fa-84e3-1463eb351fc0.png)
+
+### 6. Limit a funnel conversion to a single session
+
+Proposal: In funnels, we add “Unique sessions” as an aggregation option. Setting this value would enforce that the sequence of events must occur within a specific session.
+
+![image](https://user-images.githubusercontent.com/4813045/171521401-3b7a234c-24b4-45cd-a3d0-15ab4243efb0.png)
+
+
+### 7. Limit a path conversion to a single session
+
+Proposal: I’d consider this to be a lower priority, but we should update the logic in paths to respect the concept of a session. Today, it uses a 30 min gap between events to split paths - which is essentially the same as sessions, but subtly different.
+
+One issue that we will have to figure out is being backward compatible to data that was generated prior to session support.
+
+## Work breakout:
+
+To accomplish the above, we would need to do the following work:
+
+- Trends:
+    - Add “unique sessions” as an event aggregation option in trends
+    - Add the concept of “session properties”, and specifically the “session length” property to the taxonomic filter.
+    - Allow for session properties to be used as a property aggregation in trends.
+    - Allow for session properties to be used in breakdowns
+    - Add the ability to “auto bucket” when breaking down by numeric properties
+    - Note: while not included in the requirements above, it’d be cool to allow filtering based on session length everywhere. (e.g. total count of pageviews where the corresponding session length > 1 min)
+- Funnels
+    - Add sessions as a aggregation target
+- Paths
+    - Use sessions instead of the 30 min gap (need to consider backward compatibility here)
+
+## Questions + notes
+
+1. Do we need a sessions table?
+    1. If we add session length as a property option, those queries will have to generate a list of session durations on the fly. This query can’t be limited to specific event types because it has to query across all events to ensure it finds the first and last event in the session. As a result it might be a hefty query. To optimize this, we could create a sessions table in clickhouse (probably using AggregatingMergeTree) ****that contains the session_id and it’s length
+    2. My vote is that we don’t over optimize early and try to do without this table at first. We can always add it later.
+2. There will be a mismatch in recording duration and session duration
```

Comment:
hmmm, tricky! "Session" Recordings sounds like a misnomer now 😂. Trying to reconcile these two when they're going to be completely different sounds like a losing battle...

So, question: Do users find it valuable to align these? What problem does this solve? Or is it more of a consistency issue?

Reply:
I think there's value in the continuity of analysis. It seems that we're grafting in sessions to insights so that analysis from insights to sessions is super smooth and we're choosing that priority over session aggregates to recordings. Ideally we will allow for both ways of analysis, if you're on recordings and checking out specific sessions with some characteristics, it'd be awesome to then be able to ask "how many sessions like this have happened over time" directly from recordings. Although, maybe the time alignment isn't the biggest deal here

Reply:
> Although, maybe the time alignment isn't the biggest deal here

I agree with this. Also, with PageView and PageLeave events, these two durations should be very close to each other most of the time.

Reply:
We call them "recordings" now so I believe a note in documentation somewhere explaining this is ~fine.